### PR TITLE
Confidence Dataset

### PR DESCRIPTION
**Pull request by vratins**
A wrapper dataset that builds a dataset for the confidence model using the sampled cache created from the trained flow generator + tests. Built on top of confidence model PR.

## Summary by CodeRabbit

* **New Features**
  * Added confidence dataset support that combines candidate positions with ground-truth water structures.
  * Added configurable confidence labels, distance thresholds, candidate limits, and handling for missing or empty data.
  * Added oxygen features and metadata for ground-truth positions and nearest-water relationships.

* **Tests**
  * Added comprehensive coverage for confidence calculations, labeling modes, filtering, subsampling, validation, and edge cases.

### CODE REVIEW

File: src/confidence_dataset.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        The nearest-GT distance / argmin is computed once and reused for: the
+        regression target (soft smootherstep, or a hard 1[d<=accept_radius] when
+        `hard_label`), the AUC-PR label (1[d<=accept_radius]), and the
+        per-candidate nearest-GT index (for the optional coverage loss, which
+        groups candidates by the GT site they could cover).
+        """
+        if candidate_pos.numel() == 0 or gt_pos.numel() == 0:
+            empty = candidate_pos.new_empty(0)
+            return empty, empty, candidate_pos.new_empty(0, dtype=torch.long)
+        d, gt_index = torch.cdist(candidate_pos, gt_pos).min(dim=1)  # (Nc,), (Nc,)
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/confidence_dataset.py:169]: Fail already exists for this, we skip it

File: tests/test_confidence_dataset.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        # A structure whose waters were all filtered out. Targets come back
+        # empty while num_nodes stays at the candidate count -- such an item
+        # cannot be collated, so the trainer must exclude these structures.
+        flow = _FakeFlowDataset(["k"], gt_by_key={"k": torch.empty(0, 3)})
+        _write_candidate(tmp_path / "k.pt", torch.randn(3, 3))
+        sample = ConfidenceDataset(flow, candidate_dir=tmp_path)[0]
+        assert sample["water"].num_nodes == 3
+        assert sample["water"].target_confidence.shape == (0,)
+        assert sample["water"].label_1A.shape == (0,)
+        assert sample["water"].gt_index.shape == (0,)
```
[the remark this answers is not in the document]
Comment (vratins) [reply tests/test_confidence_dataset.py:224]: Same as ^

File: src/confidence_dataset.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        if not self.candidate_dir.exists():
+            raise FileNotFoundError(
+                f"Candidate directory not found: {self.candidate_dir}. "
+                "Build it with scripts/cache_candidates.py first."
+            )
+        self.r_in = float(r_in)
+        self.r_out = float(r_out)
+        self.hard_label = bool(hard_label)
+        self.accept_radius = float(accept_radius)
+        self.max_candidates = max_candidates
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/confidence_dataset.py:104]: addressed

File: tests/test_confidence_dataset.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
+"""Unit tests for src/confidence_dataset.py -- candidate join and target computation."""
+
+import pytest
+import torch
+import torch.nn.functional as F
+from torch_geometric.data import HeteroData
+
+from src.confidence import smootherstep_target
+from src.confidence_dataset import _oxygen_features, ConfidenceDataset
+from src.constants import EDGE_PP, NUM_RBF
```
[the remark this answers is not in the document]
Comment (vratins) [reply tests/test_confidence_dataset.py]: addressed

File: tests/test_confidence_dataset.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+        ds = ConfidenceDataset(flow, candidate_dir=tmp_path, max_candidates=5)
+        sample = ds[0]
+        assert sample["water"].num_nodes == 5
+        assert sample["water"].target_confidence.shape == (5,)
+        # a subset of the candidate cloud, not a truncation
+        rows = {tuple(r.tolist()) for r in sample["water"].pos}
+        assert rows.issubset({tuple(r.tolist()) for r in cand})
+
+        # a fresh draw each epoch, so repeated reads differ
+        assert not torch.equal(ds[0]["water"].pos, ds[0]["water"].pos)
```
[the remark this answers is not in the document]
Comment (vratins) [reply tests/test_confidence_dataset.py]: addressed

File: src/confidence_dataset.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+from torch.utils.data import Dataset
+from torch_geometric.data import HeteroData
+
+from src.confidence import smootherstep_confidence
+from src.constants import ELEM_IDX, ELEMENT_VOCAB
+from src.dataset import ProteinWaterDataset
+
+
+WATER_FEATURE_DIM = len(ELEMENT_VOCAB) + 1  # matches element_onehot in src/dataset.py
+OXYGEN_INDEX = ELEM_IDX["O"]
```
Comment (DorisMai) [inline src/confidence_dataset.py:33]: seems not specific to this file. WATER_FEATURE_DIM is identical to NODE_FEATURE_DIM. And ELEM_IDX["O"] also used in flow.py. Try to be consistent and re-use global constants.

File: src/confidence_dataset.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+
+def _oxygen_features(n: int, device: torch.device | None = None) -> Tensor:
```
Comment (DorisMai) [inline src/confidence_dataset.py:36]: again seems a general function (same oxygen onehot done twice in `flow.py`) that should live elsewhere. You might want to do a separate small PR that moves the generic `element_onehot` from `dataset.py` to `utils.py`, move this function as `oxygen_onehot` there too, and then update the imports here and in flow.py.

Also should reuse constant for `TestOxygenFeatures` (which should again probably live under a different test file).

File: src/confidence_dataset.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        The nearest-GT distance / argmin is computed once and reused for: the
+        regression target (soft smootherstep, or a hard 1[d<=accept_radius] when
+        `hard_label`), the AUC-PR label (1[d<=accept_radius]), and the
+        per-candidate nearest-GT index (for the optional coverage loss, which
+        groups candidates by the GT site they could cover).
+        """
+        if candidate_pos.numel() == 0 or gt_pos.numel() == 0:
+            empty = candidate_pos.new_empty(0)
+            return empty, empty, candidate_pos.new_empty(0, dtype=torch.long)
+        d, gt_index = torch.cdist(candidate_pos, gt_pos).min(dim=1)  # (Nc,), (Nc,)
```
Comment (DorisMai) [inline src/confidence_dataset.py:169]: this function seems largely doing the same thing as `confidence.py:_nearest_gt_distance()`. Why not make `torch.cdist` a common helper function for both places (throw away the gt_index if unused)? It seems that you even tested that the two implementations are the same in `test_default_target_is_smootherstep`. Is there a reason to keep two separate implementation? If you do decide to stick with cdist, watch out for compute_mode when the matrix is large if you can about high numerical precision.

File: src/confidence_dataset.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        # epoch covers all candidates over training.
+        if (
+            self.max_candidates is not None
+            and candidate_pos.size(0) > self.max_candidates
+        ):
+            sel = torch.randperm(candidate_pos.size(0))[: self.max_candidates]
+            candidate_pos = candidate_pos[sel]
+        n_cand = candidate_pos.size(0)
+
+        target, label_1A, gt_index = self._compute_targets(candidate_pos, gt_pos)
```
Comment (DorisMai) [inline src/confidence_dataset.py:198]: `label_1A` can be named better...can be as plain as `within_accept_radius`. Also isn't this effectively a bool then? why float?

File: src/confidence_dataset.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+        return target, label, gt_index
+
+    def __getitem__(self, idx: int) -> HeteroData:
+        flow_idx = self._indices[idx]
+        data = self.flow_dataset[flow_idx]
+
+        gt_pos: Tensor = data["water"].pos.float().clone()
+        candidate_pos: Tensor = torch.load(
+            self._paths[idx], map_location="cpu", weights_only=False
+        )["candidate_pos"].float()
```
[the remark this answers is not in the document]
Comment (DorisMai) [reply src/confidence_dataset.py:185]: @vratins

File: src/confidence_dataset.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+        self.r_in = float(r_in)
+        self.r_out = float(r_out)
+        self.hard_label = bool(hard_label)
+        self.accept_radius = float(accept_radius)
+        self.max_candidates = max_candidates
+        if self.r_out <= self.r_in:
+            raise ValueError("r_out must exceed r_in.")
+        if self.accept_radius < 0:
+            raise ValueError("accept_radius must be non-negative.")
+        if self.max_candidates is not None and self.max_candidates < 0:
```
Comment (DorisMai) [inline src/confidence_dataset.py:109]: might need to be positive, otherwise index error at line 195 (unless you enforce non-empty candidate water)

File: src/confidence_dataset.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+        # Swap GT water nodes for the candidates to be scored.
+        data["water"].pos = candidate_pos
+        data["water"].x = _oxygen_features(n_cand, device=candidate_pos.device)
+        data["water"].num_nodes = n_cand
+        data["water"].target_confidence = target
+        data["water"].label_1A = label_1A
+        data["water"].gt_index = gt_index
+        data.n_gt = torch.tensor([gt_pos.size(0)], dtype=torch.long)
+        data["water"].gt_pos = gt_pos
```
Comment (DorisMai) [inline src/confidence_dataset.py:208]: are these gt_index and gt_pos used anywhere? docstring of `_compute_targets` said they are reused -- would that be upcoming PRs? If so, please add tests on batch > 1. I am a bit concerned about the batching behavior (offset issue) here. Otherwise just remove them.

File: tests/test_confidence_dataset.py
Comment (DorisMai) [inline tests/test_confidence_dataset.py:1]: please add test under batching

**Review by DorisMai — APPROVED** [review]
Main comment is regarding batching. Other comments are mostly small refactoring.

File: src/confidence_dataset.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (vratins) [reply src/confidence_dataset.py:185]: Done

File: src/confidence_dataset.py (hunk 8 above)
Comment (vratins) [reply src/confidence_dataset.py:169]: Consolidated.

File: src/confidence_dataset.py (hunk 9 above)
Comment (vratins) [reply src/confidence_dataset.py:198]: Renamed to within_accept_radius. Kept float since it's the AUC-PR/BCE target, which needs float.

File: src/confidence_dataset.py (hunk 11 above)
Comment (vratins) [reply src/confidence_dataset.py:109]: Done

File: tests/test_confidence_dataset.py
Comment (vratins) [reply tests/test_confidence_dataset.py:1]: Added `test_items_batch_via_pyg`

File: src/confidence_dataset.py (hunk 12 above)
Comment (vratins) [reply src/confidence_dataset.py:208]: Removed them from this PR, will add them in the next PR with the confidence trainer code.

File: src/confidence_dataset.py (hunk 6 above)
Comment (vratins) [reply src/confidence_dataset.py:33]: Understood, will shift in a future PR.

File: src/confidence_dataset.py (hunk 7 above)
Comment (vratins) [reply src/confidence_dataset.py:36]: Understood, will do so in future PR.
